contracts/credits/src/contract.rs

```diff
@@ -0,0 +1,1438 @@
+        config.airdrop_address = Some(deps.api.addr_validate(&addr)?);
+    }
+    if let Some(addr) = msg.lockdrop_address {
+        config.lockdrop_address = Some(deps.api.addr_validate(&addr)?);
+    }
+    CONFIG.save(deps.storage, &config)?;
+
+    // store token info
+    let info = Cw20State::TokenInfo {
+        name: TOKEN_NAME.to_string(),
+        symbol: TOKEN_SYMBOL.to_string(),
+        decimals: TOKEN_DECIMALS,
+        total_supply: Uint128::zero(),
+        mint: Some(Cw20State::MinterData {
+            minter: config.dao_address,
+            cap: None,
+        }),
+    };
+    TOKEN_INFO.save(deps.storage, &info, env.block.height)?;
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, Cw20ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig {
+            airdrop_address,
+            lockdrop_address,
+        } => execute_update_config(deps, env, info, airdrop_address, lockdrop_address),
+        ExecuteMsg::AddVesting {
+            address,
+            amount,
+            start_time,
+            duration,
+        } => execute_add_vesting(deps, env, info, address, amount, start_time, duration),
+        ExecuteMsg::Transfer { recipient, amount } => {
+            execute_transfer(deps, env, info, recipient, amount)
+        }
+        ExecuteMsg::Withdraw {} => execute_withdraw(deps, env, info),
+        ExecuteMsg::Burn { amount } => execute_burn(deps, env, info, amount),
+        ExecuteMsg::BurnFrom { owner, amount } => execute_burn_from(deps, env, info, owner, amount),
+        ExecuteMsg::Mint {} => execute_mint(deps, env, info),
+    }
+}
+
+#[entry_point]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    Ok(Response::default())
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    airdrop_address: String,
+    lockdrop_address: String,
+) -> Result<Response, Cw20ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+    if info.sender != config.dao_address {
+        return Err(Cw20ContractError::Unauthorized {});
+    }
+
+    config.airdrop_address = Some(deps.api.addr_validate(&airdrop_address)?);
+    config.lockdrop_address = Some(deps.api.addr_validate(&lockdrop_address)?);
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default())
+}
+
+pub fn execute_add_vesting(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    address: String,
+    amount: Uint128,
+    start_time: u64,
+    duration: u64,
+) -> Result<Response, Cw20ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender
+        != config
+            .airdrop_address
+            .ok_or_else(|| StdError::generic_err("uninitialized"))?
```

Comment:
please move all inline defined errors to a ContractError enum in a separate file as we usually do

Reply:
hm, not so easy cause I used cw20 contract error, do we really need it?

Reply:
you can make errors transparent as it's done for example here with StdError and PauseError https://github.com/neutron-org/neutron-dao/blob/neutron_audit_informal_17_01_2023/contracts/tokenomics/distribution/src/error.rs

having errors as constants helps avoid errors duplication and subtle difference between similar errors as well as helps to have all possible errors structured in a list